synthetic_data/dataset_generators/datetime_generator.py

```diff
@@ -0,0 +1,84 @@
+import numpy as np
+import pandas as pd
+from numpy.random import Generator
+from typing import Optional
+
+
+def generate_datetime(
+    rng: Generator,
+    date_format: str,
+    start_date: Optional[str] = None,
+    end_date: Optional[str] = None,
+) -> str:
+    """
+    Generate datetime given the random_state, date_format, and start/end dates.
+
+    :param rng: the np rng object used to generate random values
+    :type rng: numpy Generator
+    :param date_format: the format that the generated datatime will follow,
+        defaults to None
+    :type date_format: str, None, optional
+    :param start_date: the earliest date that datetimes can be generated at,
+        defaults to None
+    :type start_date: pd.Timestamp, None, optional
+    :param start_date: the latest date that datetimes can be generated at,
+        defaults to None
+    :type start_date: pd.Timestamp, None, optional
+
+    :return: generated datetime
+    :rtype: str
+    """
+    if not start_date:
+        # 100 years in past
+        start_date = pd.Timestamp(1920, 1, 1)
+    if not end_date:
+        # protection of 30 years in future
+        end_date = pd.Timestamp(2049, 12, 31)
+    t = rng.random()
+    ptime = start_date + t * (end_date - start_date)
+
+    return ptime.strftime(date_format)
+
+
+def random_datetimes(
+    rng: Generator,
+    date_format_list: Optional[str] = None,
+    start_date: Optional[str] = None,
+    end_date: Optional[str] = None,
+    num_rows: int = 1,
+) -> np.array:
+    """
+    Generate datetime given the random_state, date_format, and start/end dates.
+
+    :param rng: the np rng object used to generate random values
+    :type rng: numpy Generator
+    :param date_format: the format that the generated datatime will follow,
+        defaults to None
+    :type date_format: str, None, optional
+    :param start_date: the earliest date that datetimes can be generated at,
+        defaults to None
+    :type start_date: pd.Timestamp, None, optional
+    :param start_date: the latest date that datetimes can be generated at,
+        defaults to None
+    :type start_date: pd.Timestamp, None, optional
+
+    :return: array of generated datetimes
+    :rtype: numpy array
+    """
+    date_list = []
+    if not date_format_list:
+        date_format_list = ["%B %d %Y %H:%M:%S"]
+
+    for _ in range(num_rows):
+        date_format = rng.choice(date_format_list)
+        datetime = generate_datetime(
+            rng, date_format=date_format, start_date=start_date, end_date=end_date
+        )
+        date_list.append(datetime)
```

Comment:
I wonder if we can pre-allocate a list to reduce the expense of the append.
something like:

```python
date_list = ["%B %d %Y %H:%M:%S"] * num_rows

and then below, use the index to insert in the correct place.
```

There might even be more interesting optimizations that could make this overall faster instead of looping each row.
But we can optimize that in a future PR if desired.